### DISCUSSION TITLE
Upgrading third-party rpm

### DISCUSSION THREAD

**Discussion by dgilbert101** [oamg/leapp-repository#1225] [Q&A] [answered]
I have a unsigned third-party rpm that I need to have upgraded during the leapp upgrade process. Its still unclear how this works with leapp. If I specify the package name in to_install the dnf_transaction_check fails with "none of the providers can be installed" error. Even though required package and its dependencies are present in the AppStream repo.  Do I need to spell out all the rpm dependencies? I would've thought that its dependencies would be automatically resolved. Any insight or help would be greatly appreciated.

**Comment by pirat89** [comment]
Hi @dgilbert101, I have pinned the previous discussion about third party packages in the discussion to make it more visible (and updated the title to hit key words). I realized that closed discussions are not visible by default, so hopefully this could help. I expect that the discussion answers your questions, but feel free to ask if not. I am keeping this opened, waiting for your feedback.
The mentioned discussion is here:
* https://github.com/oamg/leapp-repository/discussions/1192

To highligh some important facts:
* /etc/leapp/transaction/to* files exect really just package name (nothing more) specified
* to_remove, to_keep files works just for installed RH signed packages - and to_keep has little different meaning than people usually thinks
* content from to_install file is just given to DNF, in case that no entity of such name is discovered by DNF, we print warning msg and skip the requirement.
* what remains, depependencies are automatically resolved (like when you `dnf install leapp-upgrade` and all other packages are automatically installed as dependencies); just in case of RHEL in-place upgrade RPM transaction the complexity is higher (due to the size of the transaction and amount of changes between systems) so it's easier to hit conflicts here in comparison to small minor updates.
  * so, usually you do not need to spell out all dependencies, but it depends also how third-party packages are built. the answer is little bit more complex and I tried to cover it in the previous discussion.

**Reply by dgilbert101** [reply]
Thanks for the information. I tried the option with the env variables configured, and it did not make a difference. I have some epel package dependencies that need to be upgraded, along with my unsigned packages. I can perform a new install with them just fine, leapp is just being problematic. Or at least my understanding of it is causing me issues. I'm open to custom actor, what ever works.

**Comment by dgilbert101** [comment, marked as answer]
I was able to find a solution by create a custom actor that would install/upgrade my desired unsigned application.

```
from leapp.actors import Actor
from leapp.libraries.common.rpms import has_package
from leapp.libraries.stdlib import api
from leapp.models import InstalledRPM, RpmTransactionTasks
from leapp.tags import ChecksPhaseTag, IPUWorkflowTag

PRODUCT_PKGS = ["pkg1", "pkg2", "pkg3"]

class PkgUpgradeCheck(Actor):
    # Checking if pkg is installed.

    name = 'check_pkg'
    consumes = (InstalledRPM,)
    produces = (RpmTransactionTasks,)
    tags = (ChecksPhaseTag, IPUWorkflowTag,)

    def process(self):
        for PACKAGE in PRODUCT_PKGS:
            if has_package(InstalledRPM, PACKAGE):
                api.produce(RpmTransactionTasks(
                    to_install=[PACKAGE, "uriparser"]))
```

I tried to configure to_remove in the actor to remove openssl11-libs, but it would fail to properly remove it unless `LEAPP_DEVEL_RPMS_ALL_SIGNED=1 LEAPP_UNSUPPORTED=1` was used. Instead, to deal with the epel package that was causing issues, I leveraged bundled-rpm meta package that would obsolete the existing openssl11-libs pacakge. After this, was able to successfully upgrade the os and and all third-party application packages.

**Reply by pirat89** [reply]
Thanks for the sharing of your solution.
